packages/solidity-contracts/contracts/fuelchain/v2/FuelMessagePortalV2.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.9;
+
+import '../FuelMessagePortal.sol';
+
+contract FuelMessagePortalV2 is FuelMessagePortal {
+
+    error AccountDepositLimit();
+    error GlobalDepositLimit();
+
+    uint public depositLimitGlobal;
+    uint public depositLimitPerAccount;
+
+    uint public totalDeposited;
+    mapping(address => uint) public depositedAmounts;
+
+
+    function initializeV2(
+        uint _depositLimitGlobal, 
+        uint _depositLimitPerAccount
+    ) public reinitializer(2) {
+        depositLimitGlobal = _depositLimitGlobal;
+        depositLimitPerAccount = _depositLimitPerAccount;
+    }
+
+    function setGlobalDepositLimit(uint limit) external virtual payable onlyRole(DEFAULT_ADMIN_ROLE) {
+        depositLimitGlobal = limit;
+    }
+
+    function setPerAccountDepositLimit(uint limit) external virtual payable onlyRole(DEFAULT_ADMIN_ROLE) {
+        depositLimitPerAccount = limit;
+    }
+
+    function rescueETH(uint amount) external virtual payable onlyRole(DEFAULT_ADMIN_ROLE) {
+        payable(msg.sender).transfer(amount);
+    }
+
+    ////////////////////////
+    // Internal Functions //
+    ////////////////////////
+
+    /// @notice Performs all necessary logic to send a message to a target on Fuel
+    /// @param recipient The message receiver address or predicate root
+    /// @param data The message data to be sent to the receiver
+    function _sendOutgoingMessage(bytes32 recipient, bytes memory data) internal virtual override {
+        bytes32 sender = bytes32(uint256(uint160(msg.sender)));
+        unchecked {
+            //make sure data size is not too large
+            if (data.length >= MAX_MESSAGE_DATA_SIZE) revert MessageDataTooLarge();
+
+            // v2: increase deposited amount for sender
+            // Do not think this needs an overflow check, you cannot get that much ether
+            uint256 userDepositedAmount = depositedAmounts[msg.sender] + msg.value;
+            if(userDepositedAmount > depositLimitPerAccount) {
+                revert AccountDepositLimit();
+            }
+
+            depositedAmounts[msg.sender] = userDepositedAmount;
+
+            // v2: increase global deposited ether
+            uint256 globalDepositedAmount = totalDeposited += msg.value;
+            if(globalDepositedAmount > depositLimitGlobal) {
+                revert GlobalDepositLimit();
+            }
+            
+            //make sure amount fits into the Fuel base asset decimal level
+            uint256 amount = msg.value / PRECISION;
+            if (msg.value > 0) {
+                if (amount * PRECISION != msg.value) revert AmountPrecisionIncompatibility();
+                if (amount > type(uint64).max) revert AmountTooBig();
+            }
+
+
```

Comment:
```suggestion
            // v2: increase global deposited ether
            uint256 globalDepositedAmount = totalDeposited += msg.value;
            if(globalDepositedAmount > depositLimitGlobal) {
                revert GlobalDepositLimit();
            }
            
            //make sure amount fits into the Fuel base asset decimal level
            uint256 amount = msg.value / PRECISION;
            if (msg.value > 0) {
                if (amount * PRECISION != msg.value) revert AmountPrecisionIncompatibility();
                if (amount > type(uint64).max) revert AmountTooBig();
            }

            depositedAmounts[msg.sender] = userDepositedAmount;

```

Save the user some gas in case subsequent checks revert